[GStreamer] Critical warnings when browsing cnn.com

philn opened the pull request:
#### c94b3642944f7b606e8d32c196f91e2875eea54b
<pre>
[GStreamer] Critical warnings when browsing cnn.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=254622">https://bugs.webkit.org/show_bug.cgi?id=254622</a>

Reviewed by Xabier Rodriguez-Calvar.

Parsing and decoding errors are propagated from the GStreamer element harness to the ImageDecoder.
The cnn.com debug logs indicated the openh264dec was failing to negotiate caps with upstream.

This patch also includes a couple changes in the GStreamerElementHarness, mainly automatic start-up
of chained harnesses in order to avoid potentially racy behaviors in sticky events handling.

* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
(WebCore::ImageDecoderGStreamer::ImageDecoderGStreamer):
(WebCore::ImageDecoderGStreamer::encodedDataStatus const):
(WebCore::ImageDecoderGStreamer::pushEncodedData):
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.h:
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
(WebCore::GStreamerElementHarness::pushStickyEvents):
(WebCore::GStreamerElementHarness::Stream::Stream):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.h:
(WebCore::GStreamerElementHarness::isStarted const):

Canonical link: <a href="https://commits.webkit.org/263134@main">https://commits.webkit.org/263134@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9464c3b37f480aa46b7f980d5ca94d87a5bc4229

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3707 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3792 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5135 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/3989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3680 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3869 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3797 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/3212 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3751 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3343 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4968 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1484 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3312 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/3373 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3293 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/3371 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/4737 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3759 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3045 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/3291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3312 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/904 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3327 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/3568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -240,7 +243,7 @@ void ImageDecoderGStreamer::pushEncodedData(const FragmentedSharedBuffer& shared
     auto bytes = data->createGBytes();
     auto buffer = adoptGRef(gst_buffer_new_wrapped_bytes(bytes.get()));
-    m_eos = false;
+    m_eos = m_error = false;
```
Does the style checker allow this?

philn replied:
Apparently ~~not~~ :)
I was told the reported bug is not fixed by this PR though, so needs another round but I have no time for this.

calvaris replied:
I still think it is better to have it in different lines.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -94,6 +94,34 @@ ImageDecoderGStreamer::ImageDecoderGStreamer(FragmentedSharedBuffer& data, const
+    GRefPtr<GstElement> parsebin = gst_element_factory_make("parsebin", makeString("image-decoder-parser-", decoderId.exchangeAdd(1)).utf8().data());
+    m_parserHarness = GStreamerElementHarness::create(WTFMove(parsebin), [](auto&, const auto&) { }, [this](auto& pad) -> RefPtr<GStreamerElementHarness> {
+        auto caps = adoptGRef(gst_pad_query_caps(pad.get(), nullptr));
+        auto identityHarness = GStreamerElementHarness::create(adoptGRef(gst_element_factory_make("identity", nullptr)), [](auto&, const auto&) { });
```
You can't adopt a floating ref!

philn replied:
ah right, should be a normal `GRefPtr<>`

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -94,6 +94,34 @@ ImageDecoderGStreamer::ImageDecoderGStreamer(FragmentedSharedBuffer& data, const
+        auto identityHarness = GStreamerElementHarness::create(adoptGRef(gst_element_factory_make("identity", nullptr)), [](auto&, const auto&) { });
+        GST_DEBUG_OBJECT(pad.get(), "Caps on parser source pad: %" GST_PTR_FORMAT, caps.get());
+        if (!caps || !doCapsHaveType(caps.get(), "video") || m_decoderHarness) {
+            GST_WARNING_OBJECT(m_decoderHarness->element(), "Parsed stream caps refused. Decoder already configured: %s", boolForPrinting(m_decoderHarness.get()));
```
This message looks a bit confusing to understand.

philn replied:
What's confusing?

philn replied:
I'll split this in two warnings

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -94,6 +94,34 @@ ImageDecoderGStreamer::ImageDecoderGStreamer(FragmentedSharedBuffer& data, const
+        m_decoderHarness = GStreamerElementHarness::create(WTFMove(element), [this](auto& stream, const auto& outputBuffer) {
+            auto outputCaps = stream.outputCaps();
+            auto sample = adoptGRef(gst_sample_new(outputBuffer.get(), outputCaps.get(), nullptr, nullptr));
+            storeDecodedSample(WTFMove(sample));
```
I would adopt into the store function directly

philn committed:
[GStreamer] Critical warnings when browsing cnn.com
https://bugs.webkit.org/show_bug.cgi?id=254622

Reviewed by Xabier Rodriguez-Calvar.

Parsing and decoding errors are propagated from the GStreamer element harness to the ImageDecoder.
The cnn.com debug logs indicated the openh264dec was failing to negotiate caps with upstream.

This patch also includes a couple changes in the GStreamerElementHarness, mainly automatic start-up
of chained harnesses in order to avoid potentially racy behaviors in sticky events handling.

* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
(WebCore::ImageDecoderGStreamer::ImageDecoderGStreamer):
(WebCore::ImageDecoderGStreamer::encodedDataStatus const):
(WebCore::ImageDecoderGStreamer::pushEncodedData):
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.h:
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
(WebCore::GStreamerElementHarness::pushStickyEvents):
(WebCore::GStreamerElementHarness::Stream::Stream):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.h:
(WebCore::GStreamerElementHarness::isStarted const):

Canonical link: https://commits.webkit.org/263134@main
